include/xtypes/idl/parser.hpp

```diff
@@ -34,55 +34,154 @@
 #include <memory>
 #include <exception>
 #include <locale>
+#include <regex>
 
 namespace eprosima {
 namespace xtypes {
 namespace idl {
 
+class Parser;
+
+struct Context
+{
+    // Config
+    bool ignore_case = false;
+    bool clear = true;
+    bool preprocess = true;
+    std::string preprocessor_exec = "";
+    std::vector<std::string> include_paths;
+
+    // Results
+    bool success = false;
+    std::map<std::string, DynamicType::Ptr> structs;
+    std::map<std::string, DynamicData> constants;
+    //std::map<std::string, Modules> modules;
+
+    ~Context()
+    {
+        clear_context();
+    }
+
+private:
+    friend class Parser;
+    Parser* instance_;
+
+    inline void clear_context();
+};
+
+static const Context DEFAULT_CONTEXT = Context();
+
+
 class Parser
 {
 public:
+    static Parser* instance()
+    {
+        if (my_instance_ == nullptr)
+        {
+            my_instance_ = new Parser();
+        }
+        return my_instance_;
+    }
+
+    static void destroy()
+    {
+        delete my_instance_;
+        my_instance_ = nullptr;
+    }
+
     Parser()
         : parser_(idl_grammar())
     {
         parser_.enable_ast();
     }
 
+    Context parse(
+            const std::string& idl_string)
+    {
+        Context context = DEFAULT_CONTEXT;
+        parse(idl_string, context);
+        return context;
+    }
+
     bool parse(
-            const char* idl_string,
-            bool ignore_case = false)
+            const std::string& idl_string,
+            Context& context)
     {
+        context.instance_ = this;
         std::shared_ptr<peg::Ast> ast;
-        ignore_case_ = ignore_case;
-        if (!parser_.parse(idl_string, ast))
+        ignore_case_ = context.ignore_case;
+        clear_ = context.clear;
+        std::string idl_to_parse = idl_string;
+        preprocessor(context.preprocessor_exec);
+        include_paths_ = context.include_paths;
+        if (context.preprocess)
+        {
+            idl_to_parse = preprocess_string(idl_to_parse);
+        }
+        if (!parser_.parse(idl_to_parse.c_str(), ast))
         {
+            context.success = false;
             return false;
         }
         ast = peg::AstOptimizer(true).optimize(ast);
-        build_on_ast(ast)->fill_all_types(types_map_);
+        build_on_ast(ast);//->fill_all_types(types_map_);
```

Comment:
Should the comment be removed? Outdated? (see similar occurences)

Reply:
Yes, it's old code.